### DISCUSSION TITLE
How do I apply an RTM tag and release versions? MinVer consistently adds a prerelease versions to my releases

### DISCUSSION THREAD

**Discussion by jeffward01** [adamralph/minver#703] [Q&A] [answered]
Hello all,

I have reviewed the documentation deeply, and also have researched very much what a `RTM` tag is.  I cannot find on any Microsoft or MinVer docs how I can add an RTM tag. I understand a RTM tag means 'Release to Market', however, I do not know how to add one.

My use case is that I have a dotnet cli package that I need to iterate the versions:

Such as:

1.1.1
1.1.2
1.1.3

Currently its:

1.1.1-alpha.0.2
1.1.1-alpha.0.3

I understand that if I want to increment the version, I can simply:

`> git tag x.x.x` <-- New version
`> git push --tags` 

The issue is that `-alpha.0.x` is still applied.

How do I remove the pre-release tag? 

Thanks!

**Comment by adamralph** [comment, marked as answer]
> `> git tag x.x.x` <-- New version
> `> git push --tags`

That is all that is required. If that isn't working, there must be something else wrong with your setup.

It may be best if you push a project to GitHub which demonstrates your problem.

**Reply by jeffward01** [reply]
Thanks for the quick reply!  Thanks for the help, i'll commit a small test project after this last question:

----

# Usecase:
- I create many nuget library packages, these I use an autoincrementing strategy within Azure Pipelines, its fine and works for nuget packages **specifically**.
- I have a `dotnet cli tool` I have created that I have pushed to my private nuget feed.
- When my private nuget feed 'builds and deploys' the `dotnet cli tool` it is always named `<myToolName>1.0.0.nupkg`, This is a problem for me because I cannot run `dotnet tool update <myToolName> -g` because the versions will be both equal to 1.0.0 and an update action will not be performed due to the matching versions.
- I am looking for a solution to solve this specifically, so that upon each build or other event, the versions will increment explicitly such as:

`1.2.2`
`1.2.3`
`1.2.4`
`1.2.5`


MinVer DOES work fine, however, If I do not use the 'hack' I explain below, MinVer applies  `x.x.x-<pre-release>x.x` to the end if the nuget package so that it looks like this:  `<myToolName>1.0.0-<pre-release>x.x.nupkg`.

My goal is that I do not ever want: `-<pre-release>x.x` applied to the version. I give more details below. 


# Question
Is it impossible to have a version that is: `x.x.x` instead of `x.x.x-<pre-release>x.x` ? 

Currently I use a small modification (that you suggested for a different issue actually <3) like this to achieve `x.x.x`, but in doing so:

````xml
<!-- Code was found here (thank you): https://github.com/adamralph/minver/issues/146 -->
<Target Name="MyTarget" AfterTargets="MinVer" >
    <PropertyGroup>
        <PackageVersion>$(MinVerMajor).$(MinVerMinor).$(MinVerPatch)</PackageVersion>
        <Version>$(PackageVersion)</Version>
    </PropertyGroup>
</Target>
````

 I cannot auto-increment the `PatchNumber`, which is not ideal but is alright.  I have a feeling like I am doing something wrong if I cannot create `x.x.x` versions with MinVer and must always have `-<pre-release-tag>x.x` appended. Using the 'hack' above, in order to increment the versions, I must perform:

`> git tag x.x.x` <-- New version
`> git push --tags`

Which is fine and by design, my question in summary is:

**In the same way that MinVer appends the auto-increment per each commit, can MinVer auto-increment WITHOUT appending any pre-release tag?**

---


# Goal
My goal is I want to release packages without applying`-<pre-release-tag>x.x` under any circumstances. Such as:

`1.2.2`
`1.2.3`
`1.2.4`
`1.2.5`

Is that possible with MinVer? Or will MinVer always apply the  `-<pre-release-tag>x.x`?

**Reply by adamralph** [reply]
> In the same way that MinVer appends the auto-increment per each commit, can MinVer auto-increment WITHOUT appending any pre-release tag?

I'm afraid not. This is explained in detail in [_How it works_](https://github.com/adamralph/minver#how-it-works).

If you want to increment the patch number on every commit then I'm afraid MinVer is not the package you are looking for.

**Reply by jeffward01** [reply]
Thanks for the reply!  Thats very helpful.

Last follow up question, if I use MinVer, how do I 'create a release' with only `x.x.x` (without any pre-release appended).  

I ask because it seems as long as MinVer is 'active', the pre-release will always be appended.

So in the scenario that:

1. As a developer I am creating a CLI application
2.  As a developer, I want to create releases that can be consumed by 'target audience' (developers)
3.  As a developer, I want the releases to be in the format `x.x.x`
4.  As a developer, I want to release **pre-releases** in the format `x.x.x-prerelease.x.x` (this is where MinVer comes in)
5.  As a developer I want my target audience to know have the option between consuming a `release` and `pre-release`.

I think my understanding is wrong, but it seems to me that as long as MinVer is used within the project, the project cannot have a release with `x.x.x`.  I do not see how options `3` and `5` can be accomplished with MinVer installed in the project. I think I am missing something, am I?

Does this mean that by using MinVer, the project is in 'perma-prerelease' mode?

I hope this makes sense!

**Reply by adamralph** [reply]
The first two lines in [_How it works_](https://github.com/adamralph/minver#how-it-works)—

> - If the current commit has a version tag:
>   - The version is used as-is.

So just create a tag with whatever version you want, e.g.

```bash
git tag 1.2.2
```

**Reply by nietras** [reply]
I was sad to see that it is not possible to make automatic versioning without pre-release being added by MinVer, as the question here pertains to. Same issue in that we don't want to have to tag every time we want to release but instead want to auto-release patch (or revision) incremented versions. That is instead just tag with `vX.Y.0` and patch always being zero. And then increment with height after. It seems to me this would be possible if one could simply get a `MinVerHeight` property and use that directly in version. Or if not patch version then use it in revision version. 

I think it could be a bit clearer in MinVer documentation that MinVer will *always* add a pre-release version suffix to any commit that is not "tagged". Unless you define or override the versions yourself. If I understand correctly.

**Reply by adamralph** [reply]
@nietras

> we don't want to have to tag every time we want to release

That is the entire premise behind MinVer. It is a tag-and-then-release package. If you want to do something else then it's probably not the right package for you. IIRC [Nerdbank.GitVersioning](https://github.com/dotnet/Nerdbank.GitVersioning) may do something closer to what you want.

> I think it could be a bit clearer in MinVer documentation that MinVer will always add a pre-release version suffix to any commit that is not "tagged".

I think this is quite clear in [_How it works_](https://github.com/adamralph/minver#how-it-works) but please feel free to raise a PR if you find a way to make it clearer.
